src/coreclr/jit/valuenum.cpp

```diff
@@ -5126,6 +5126,43 @@ void Compiler::fgValueNumberFieldLoad(GenTree* loadTree, GenTree* baseAddr, Fiel
 {
     noway_assert(fieldSeq != nullptr);
 
+    // Check if the load represents a frozen gc object located in a struct which is stored in a static readonly field,
+    // e.g.:
+    //
+    //   struct MyStruct { string Name; }
+    //   static readonly MyStruct MyStr = new() { Name = "Hey!" };
+    //
+    //   string GetName() => MyStr.Name; // <- loadTree
+    //
+    if ((baseAddr == nullptr) && loadTree->TypeIs(TYP_REF) &&
+        // it should be a static field
+        (fieldSeq->GetKind() == FieldSeq::FieldKind::SimpleStatic) &&
+        // boxed static to be precise
```

Comment:
This is fragile condition. We are assuming that the runtime uses boxed statics for this pattern. Is there a way to write this in a more general way?

Reply:
https://github.com/dotnet/runtime/blob/c0b6956bafb8e5d5ca8c9e878a168b285ea61d97/src/coreclr/jit/valuenum.cpp#L5125-L5145

we accept loadTree that looks like this:
![image](https://user-images.githubusercontent.com/523221/223275873-af97f7bf-1539-4567-8be9-92c91a22049d.png)

and the fieldseq that represents that load. So technically I don't need to check that tree at all. @SingleAccretion @jakobbotsch could you please validate JIT part of this PR?